lib/utils.ex

```diff
@@ -1,26 +1,97 @@
 defmodule Scidata.Utils do
   @moduledoc false
-  require Logger
 
-  def unzip_cache_or_download(base_url, zip, data_path) do
-    path = Path.join(data_path, zip)
+  def get!(url) do
+    request = %{
+      url: url,
+      headers: []
+    }
 
-    data =
-      if File.exists?(path) do
-        Logger.debug("Using #{zip} from #{data_path}\n")
-        File.read!(path)
-      else
-        Logger.debug("Fetching #{zip} from #{base_url}\n")
-        :inets.start()
-        :ssl.start()
+    request
+    |> if_modified_since()
+    |> run!()
+    |> raise_errors()
+    |> decode()
+    |> handle_cache()
+    |> elem(1)
+  end
+
+  defp if_modified_since(request) do
+    case File.stat(cache_path(request)) do
+      {:ok, stat} ->
+        value = stat.mtime |> NaiveDateTime.from_erl!() |> format_http_datetime()
+        update_in(request.headers, &[{'if-modified-since', String.to_charlist(value)} | &1])
+
+      _ ->
+        request
+    end
+  end
+
+  defp format_http_datetime(datetime) do
+    Calendar.strftime(datetime, "%a, %d %b %Y %H:%m:%S GMT")
+  end
+
+  defp run!(request) do
+    http_opts = []
+    opts = [body_format: :binary]
+    arg = {request.url, request.headers}
+
+    case :httpc.request(:get, arg, http_opts, opts) do
+      {:ok, {{_, status, _}, headers, body}} ->
+        response = %{status: status, headers: headers, body: body}
+        {request, response}
+
+      {:error, reason} ->
+        raise inspect(reason)
+    end
+  end
+
+  defp raise_errors({request, response}) do
+    if response.status >= 400 do
+      raise "HTTP #{response.status} #{inspect(response.body)}"
+    else
+      {request, response}
+    end
+  end
+
+  defp decode({request, response}) do
+    cond do
+      response != 200 ->
+        {request, response}
+
+      String.ends_with?(request.url, ".tar.gz") ->
+        {:ok, files} = :erl_tar.extract({:binary, response.body}, [:memory, :compressed])
+        response = %{response | body: files}
+        {request, response}
 
-        {:ok, {_status, _response, data}} = :httpc.request(base_url ++ zip)
-        File.mkdir_p!(data_path)
-        File.write!(path, data)
+      Path.extname(request.url) == ".gz" ->
+        body = :zlib.gunzip(response.body)
+        response = %{response | body: body}
+        {request, response}
 
-        data
-      end
+      true ->
+        {request, response}
+    end
+  end
+
+  defp handle_cache({request, response}) do
+    path = cache_path(request)
+
+    if response.status == 304 do
+      # Logger.debug(["loading cached ", path])
+      response = path |> File.read!() |> :erlang.binary_to_term()
+      {request, response}
+    else
+      File.mkdir_p!(Path.dirname(path))
+      # Logger.debug(["writing cache ", path])
+      File.write!(path, :erlang.term_to_binary(response))
```

Comment:
Given we are only interested in the body, we can simplify this by:

1. Return only the response body

2. Caching the original response body as is - this means it will be compressed and take less disk space. This is specially important because some of those datasets can be really large

Reply:
Good call on storing raw body. Is it ok to still pass it through term_to_binary or it will słów it down unnecessarily? I think its important to cache and return headers when we use this function to debug and include future datasets.

Reply:
The overhead is minimal but I would avoid doing it. If you are debugging, you can debug the outcome of the HTTP request directly (i.e. the 304/200 result).

Reply:
Ok so we will return status and headers of the response we got (eg 304) but load just the body from cache if available, right? Neat idea, how come i didn’t think about it

Reply:
Yes, we are only interested on the body, so we don't care about anything else. :)